### DISCUSSION TITLE
next.js direct links are not working getting 404 page not found

### DISCUSSION THREAD

**Discussion by poojaramani219** [vercel/next.js#17632] [Help]
http://test/shop/left-sidebar if i run this URL directly to the browser 404 Not Found error is coming but if i first go to the index page and then from link go to that page it is rendering and this issue is also coming when refreshes the page my server.js code is below

const express = require('express')
const next = require('next')

const port = parseInt(process.env.PORT, 10) || 3000
const dev = process.env.NODE_ENV !== 'production'
const app = next({ dev })
const handle = app.getRequestHandler()

app.prepare().then(() => {
  const server = express()

  server.get('/element/certificates', (req, res) => {
    return app.render(req, res, '/element/certificates')
  })

  server.get('/element/product-tabs', (req, res) => {
    return app.render(req, res, '/element/product-tabs')
  })

  server.get('/element/team-members', (req, res) => {
    return app.render(req, res, '/element/team-members')
  })

  server.get('/element/testimonials', (req, res) => {
    return app.render(req, res, '/element/testimonials')
  })

  server.get('/element/videos', (req, res) => {
    return app.render(req, res, '/element/videos')
  })

  server.get('/other/about-two', (req, res) => {
    return app.render(req, res, '/other/about-two')
  })

  server.get('/other/cart', (req, res) => {
    return app.render(req, res, '/other/cart')
  })

  server.get('/other/checkout', (req, res) => {
    return app.render(req, res, '/other/checkout')
  })

  server.get('/other/contact', (req, res) => {
    return app.render(req, res, '/other/contact')
  })

  server.get('/other/login-register', (req, res) => {
    return app.render(req, res, '/other/login-register')
  })

  server.get('/other/my-account', (req, res) => {
    return app.render(req, res, '/other/my-account')
  })

  server.get('/other/wishlist', (req, res) => {
    return app.render(req, res, '/other/wishlist')
  })

  server.get('/shop/left-sidebar', (req, res) => {
    return app.render(req, res, '/shop/left-sidebar')
  })

  server.get('/shop/product-basic/:id', (req, res) => {
    const params = { id: req.params.id };
    return app.render(req, res, '/shop/product-basic', params);
  });

  server.get('*', (req, res) => {
    return handle(req, res)
  })

  server.listen(port, (err) => {
    if (err) throw err
    console.log(`> Ready on http://localhost:${port}`)
  })
})

**Comment by mxmtsk** [comment]
Please provide some code for context like the code for the link and what routes you have in the `pages` directory

**Reply by poojaramani219** [reply]
updated with server.js

**Comment by francisrod01** [comment]
@poojaramani219 one question: Why are you using Next.js custom server instead of Api pages?
API Routes: https://nextjs.org/docs/api-routes/introduction
